src/crypto/ecies.ts

```diff
@@ -0,0 +1,265 @@
+import { ec as EC } from 'elliptic'
+import { crypto as browserCrypto } from './encryption'
+const ec = new EC('secp256k1')
+
+const subtle = browserCrypto.subtle
+
+const EC_GROUP_ORDER = Buffer.from(
+  'fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141',
+  'hex'
+)
+const ZERO32 = Buffer.alloc(32, 0)
+
+export type Ecies = {
+  iv: Buffer
+  ephemeralPublicKey: Buffer
+  ciphertext: Buffer
+  mac: Buffer
+}
+
+function assert(condition: boolean, message: string) {
+  if (!condition) {
+    throw new Error(message || 'Assertion failed')
+  }
+}
+
+function isScalar(x: unknown) {
+  return Buffer.isBuffer(x) && x.length === 32
+}
+
+function isValidPrivateKey(privateKey: Buffer) {
+  if (!isScalar(privateKey)) {
+    return false
+  }
+  return (
+    privateKey.compare(ZERO32) > 0 && // > 0
+    privateKey.compare(EC_GROUP_ORDER) < 0
+  ) // < G
+}
+
+// Compare two buffers in constant time to prevent timing attacks.
+function equalConstTime(b1: Buffer, b2: Buffer) {
+  if (b1.length !== b2.length) {
+    return false
+  }
+  let res = 0
+  for (let i = 0; i < b1.length; i++) {
+    res |= b1[i] ^ b2[i] // jshint ignore:line
+  }
+  return res === 0
+}
+
+/* This must check if we're in the browser or
+not, since the functions are different and does
+not convert using browserify */
+function randomBytes(size: number): Buffer {
+  const arr = new Uint8Array(size)
+  browserCrypto.getRandomValues(arr)
+  return Buffer.from(arr)
+}
+
+async function sha512(msg: Buffer) {
+  const digest = await subtle.digest('SHA-512', msg)
+  return Buffer.from(digest)
+}
+
+function getAes(
+  op: 'encrypt' | 'decrypt'
+): (iv: Buffer, key: Buffer, data: Buffer) => Promise<Buffer> {
+  return function (iv: Buffer, key: Uint8Array, data: Uint8Array) {
+    return new Promise(function (resolve) {
+      const importAlgorithm = { name: 'AES-CBC' }
+      const keyp = subtle.importKey('raw', key, importAlgorithm, false, [op])
+      return keyp
+        .then(function (cryptoKey) {
+          const encAlgorithm = { name: 'AES-CBC', iv }
+          return subtle[op](encAlgorithm, cryptoKey, data)
+        })
+        .then(function (result) {
+          resolve(Buffer.from(new Uint8Array(result)))
+        })
+    })
+  }
+}
+
+const aesCbcEncrypt = getAes('encrypt')
+const aesCbcDecrypt = getAes('decrypt')
+
+async function hmacSha256Sign(key: Buffer, msg: Buffer) {
+  const newKey = await subtle.importKey(
+    'raw',
+    key,
+    { name: 'HMAC', hash: { name: 'SHA-256' } },
+    false,
+    ['sign']
+  )
+
+  return Buffer.from(
+    await subtle.sign({ name: 'HMAC', hash: 'SHA-256' }, newKey, msg)
+  )
+}
+
+// function hmacSha256Sign(key, msg) {
+//   return new Promise(function (resolve) {
+//     const hmac = nodeCrypto.createHmac('sha256', Buffer.from(key))
+//     hmac.update(msg)
+//     const result = hmac.digest()
+//     resolve(result)
+//   })
+// }
+
+async function hmacSha256Verify(key: Buffer, msg: Buffer, sig: Buffer) {
```

Comment:
Replaces [this](https://github.com/bitchan/eccrypto/blob/master/browser.js#L108)